### ISSUE TITLE
[clang] -Wunreachable-code does not work on macros

### ISSUE THREAD

**Issue by tiagomacarios** [llvm/llvm-project#116296]
For the code below clang should emit 2 unreachable code warnings, but it only emits one

https://godbolt.org/z/fT9rs5xYn
```c++
#include <type_traits>

[[noreturn]] void crash() noexcept;

#define VerifyElseCrash(b) \
    if (!b) [[unlikely]] { \
        crash();           \
    }

void some_other_funtion();

int f1() {
    if (!false) [[unlikely]] {
        crash();
    }

    some_other_funtion(); // dead code
}

int f2() {
    VerifyElseCrash(false);

    some_other_funtion(); // dead code
}
```

**Comment by AaronBallman** [comment]
I believe this is behaving by design -- we often suppress diagnostics if they're generated from a macro because conditional compilation may cause that macro to trigger a diagnostic in one configuration but be fine in another configuration. See 6d9bb56cd35f865662755936977f4bc691289e2b

However, the expectation was that the analysis would get smarter over time about what is and isn't a configuration value, so there is still room for improvement here, thus the bug is being left open but marked as a QoI enhancement.

**Comment by tiagomacarios** [comment]
Yeah I saw that code, but I don't fully understand the intent. Can you give an example?

The analysis is done at the AST level, so the macro has already been fully expanded. Why would other "configurations" matter?

**Comment by AaronBallman** [comment]
Consider something like:
```
#if defined(CRASH_ON_INVALID)
#define VerifyElseCrash(b) \
    if (!b) [[unlikely]] { \
        crash();           \
    }
#else
#define VerifyElseCrash(b) \
    if (!b) [[unlikely]] { \
        log(__FILE__, __LINE__);           \
    }
#endif
```
where having `CRASH_ON_INVALID` defined means the code is actually unreachable, otherwise the code is reachable.
